### PR TITLE
FiBo-303: Migrate normalized EER to Django

### PR DESCRIPTION
**Pull request by jens1o**

### CODE REVIEW

File: backend/api/views.py [hunk 1]
```diff
@@ -64,9 +65,11 @@ def post(self, request):
             cashflow_type = request.data['type']
             overall_value = request.data['overallValue']
             timestamp = request.data['timestamp']
-            category, _ = Category.objects.get_or_create(name=request.data['category'])
+            category, _ = Category.objects.get_or_create(
+                name=request.data['category'])
             place = request.data['place']
-            place_address, _ = Place.objects.get_or_create(address=place['address'], name=place['name'])
+            place_address, _ = Store.objects.get_or_create(
```
Comment (jens1o) [inline backend/api/views.py:71]: fixme ... should not only be `Store` but `Private` also

File: backend/api/serializers.py [hunk 2]
```diff
@@ -14,9 +14,9 @@ class Meta:
         fields = '__all__'
 
 
-class PlaceSerializer(ModelSerializer):
+class StoreSerializer(ModelSerializer):
     class Meta:
-        model = Place
+        model = Store
         fields = '__all__'
```
Comment (Jonas-Karl) [inline backend/api/serializers.py:20]: Why is it the store instead of the source? In generell I thought the file should contain not only FiBoUser but also LiteUser and PremiumUser. The same I expected for the SourceSpecializations. (probably I got unsufficient knowledge about it :D)

**Review by Jonas-Karl — CHANGES_REQUESTED** [review]
I focused only on the models.py yet

File: backend/api/models.py [hunk 3]
```diff
@@ -2,38 +2,67 @@
...
 class Cashflow(models.Model):
     is_income = models.BooleanField()
     overall_value = models.DecimalField(max_digits=6, decimal_places=2)
     created = models.DateTimeField(auto_now_add=True)
     updated = models.DateTimeField(auto_now=True)
-    category = models.ForeignKey(Category, on_delete=models.DO_NOTHING)
-    place = models.ForeignKey(Place, on_delete=models.DO_NOTHING, blank=True, null=True)
+    category = models.ForeignKey(Category, on_delete=models.CASCADE)
+    source = models.ForeignKey(
+        Source, on_delete=models.SET_NULL, blank=True, null=True)
```
[the remark this answers is not in the document]
Comment (jens1o) [reply backend/api/models.py:46]: This is not how it works, is it? Here I declare that whenever a Cashflow is deleted, the Source should be kept and not deleted as well. So, in the future, you can easily reselect the place, because you had a cashflow there in the past.

File: backend/api/serializers.py (hunk 2 above)
Comment (jens1o) [reply backend/api/serializers.py:20]: Ah, I see. This is true, I guess that this will automatically be the case when all tests are passing again.

File: backend/api/models.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Jonas-Karl) [reply backend/api/models.py:46]: I think it means, whenever a Source will be deleted, the cashflow will still be existent. It will then have a source = null value stored. As visible in the EER, it is not allowed to have an optional store.

[the remark this answers is not in the document]
Comment (jens1o) [reply backend/api/models.py]: Thanks for the heads up, I had an error in thought. It is fixed here: https://github.com/Cebox82/FiBo/commit/ec40c89f9e236c78839f83252f86acecd45fad11

[the remark this answers is not in the document]
Comment (jens1o) [reply backend/api/models.py]: (At least I hope so! Please verify I fixed the issue as you'd expect.)

File: backend/api/views.py [hunk 4]
```diff
@@ -55,66 +60,80 @@ def post(self, request):
...
             cashflow.updated = datetime.now()
             cashflow_type = request.data['type']
-        except:
+        except AttributeError as e:
+            print(e.__cause__)
+            return JsonResponse({'success': False, 'message': 'Invalid source type'},
+                                status=status.HTTP_400_BAD_REQUEST)
+        except Exception as e:
+            # Fixme differentiate between other exception types
+            print(e.__cause__)
```
Comment (jens1o) [inline backend/api/views.py:136]: do you want to tackle this in this PR?

**Comment by jens1o** [comment]
very nice overall! I only checked the diff, but it looks nice so far!

File: backend/api/views.py (hunk 4 above)
Comment (Jonas-Karl) [reply backend/api/views.py:136]: No I think this will be too much and should be done in a seperate ticket

**Review by ChrisEwert — APPROVED** [review]
Looks good

**Review by jens1o — COMMENTED** [review]
is the documentation supposed to be updated in this PR?

File: backend/api/views.py [hunk 5]
```diff
@@ -130,99 +145,136 @@ def put(self, request, cashflow_id):
...
+            return JsonResponse({'success': False}, status=status.HTTP_404_NOT_FOUND)
+
+        serializer = PrivateSerializer(private, many=False)
+        return Response(serializer.data, status=status.HTTP_200_OK)
+
+
+class ItemView(APIView):
+    # permission_classes = (IsAuthenticated,)
```
Comment (jens1o) [inline backend/api/views.py:254]: is this by intent?

Comment (Jonas-Karl) [reply backend/api/views.py:254]: No this is an issue good point!

File: backend/api/views.py [hunk 6]
```diff
@@ -130,99 +145,136 @@ def put(self, request, cashflow_id):
...
+
+    def post(self, request, cashflow_id):
+        try:
```
Comment (jens1o) [inline backend/api/views.py:257]: Please add a FIXME that authorization should be checked (i.e. whether the user may access this cashflow by managing the corresponding account).

File: backend/api/views.py [hunk 7]
```diff
@@ -130,99 +145,136 @@ def put(self, request, cashflow_id):
...
+                                       cashflow=cashflow)
+        except Exception as e:
+            # Fixme differentiate between other exception types
+            print(e.__cause__)
+            return JsonResponse({'success': False}, status=status.HTTP_400_BAD_REQUEST)
+
+        return JsonResponse({'success': True, 'item': item.id}, status=status.HTTP_201_CREATED)
+
+    def get(self, request, cashflow_id, item_id):
+        try:
```
Comment (jens1o) [inline backend/api/views.py:270]: Please add a FIXME here, as well.

File: backend/api/views.py [hunk 8]
```diff
@@ -130,99 +145,136 @@ def put(self, request, cashflow_id):
...
+            item = Item.objects.get(id=item_id)
+            if item.cashflow.id != cashflow_id:
+                raise AttributeError('invalid item and cashflow combination')
+        except Exception as e:
+            # FIXME: Differentiate between ObjectDoesNotExist and a broad Exception
```
Comment (jens1o) [inline backend/api/views.py:275]: should this be tackled in this PR?

Comment (Jonas-Karl) [reply backend/api/views.py:275]: No, the FIXME should not be tackled in this PR.

In case if you meant the following lines:

```python
 if item.cashflow.id != cashflow_id:
                raise AttributeError('invalid item and cashflow combination')
```

I think it is not wrong to add this exception there as I had to create this whole ItemView class in order to support items and it is also already tested in test_views. Are you worried about reducing the size of this PR or an out of scope action?

File: backend/api/views.py (hunk 7 above)
Comment (Jonas-Karl) [reply backend/api/views.py:274]: I added it to the class. Is it enough or does each method need a FIXME?

Comment (jens1o) [reply backend/api/views.py:274]: fine for me :)

**Review by jens1o — COMMENTED** [review]
cannot approve, but lgtm. I think @XSnelliusX wanted to review, still :)

**Review by XSnelliusX — APPROVED** [review]
lgtm :)

**Review by XSnelliusX — APPROVED** [review]
After testing the changes on the correct branch this still looks good to me :)

File: backend/docs/API_Documentation.yaml [hunk 9]
```diff
@@ -112,173 +104,300 @@ paths:
...
+
         '401':
-          $ref: '#/components/responses/accessError'
+          $ref: '#/components/responses/access_error'
+
         '500':
-          description: The Cashflow could not be created. There are multiple reasons, for this. For example the account id form the request may be not in the database
-  /cashflow/<int:cashflow_id>:
+          description: The Cashflow could not be updated. There are multiple reasons, for this. For example the account id form the request may be not in the database
+  /cashflows/<int:cashflow_id>/:
```
Comment (Jonas-Karl) [inline backend/docs/API_Documentation.yaml:134]:
```suggestion
          description: The Cashflow could not be created. There are multiple reasons, for this. For example the account id form the request may be not in the database
  /cashflows/<int:cashflow_id>/:
```

revert this change

File: backend/docs/API_Documentation.yaml [hunk 10]
```diff
@@ -112,173 +104,300 @@ paths:
...
       security:
         - bearerAuth: [ ]
       parameters:
         - $ref: '#/components/parameters/auth'
         - $ref: '#/components/parameters/cashflow_id'
       responses:
         '200':
-          $ref: '#/components/responses/putAndDeleteCashflow.200'
+          $ref: '#/components/responses/put_and_delete_cashflow.200'
+
```
Comment (Jonas-Karl) [inline backend/docs/API_Documentation.yaml:199]:
```suggestion
```

delete line

File: backend/docs/API_Documentation.yaml [hunk 11]
```diff
@@ -112,173 +104,300 @@ paths:
...
         '401':
-          $ref: '#/components/responses/accessError'
+          $ref: '#/components/responses/access_error'
+
```
Comment (Jonas-Karl) [inline backend/docs/API_Documentation.yaml:202]:
```suggestion
```

delete line

File: backend/docs/API_Documentation.yaml [hunk 12]
```diff
@@ -112,173 +104,300 @@ paths:
...
         '500':
           $ref: '#/components/responses/cashflows.500'
-  /place/:
+
```
Comment (Jonas-Karl) [inline backend/docs/API_Documentation.yaml:205]:
```suggestion
```

delete line

File: backend/docs/API_Documentation.yaml [hunk 13]
```diff
@@ -112,173 +104,300 @@ paths:
...
         '400':
           $ref: '#/components/responses/400'
+
         '401':
-          $ref: '#/components/responses/accessError'
+          $ref: '#/components/responses/access_error'
+
         '500':
           $ref: '#/components/responses/cashflows.500'
+
```
Comment (Jonas-Karl) [inline backend/docs/API_Documentation.yaml:184]:
```suggestion
```

delete line

File: backend/docs/API_Documentation.yaml [hunk 14]
```diff
@@ -112,173 +104,300 @@ paths:
...
-          $ref: '#/components/responses/putAndDeleteCashflow.200'
+          $ref: '#/components/responses/put_and_delete_cashflow.200'
+
         '400':
           $ref: '#/components/responses/400'
+
         '401':
-          $ref: '#/components/responses/accessError'
+          $ref: '#/components/responses/access_error'
+
```
Comment (Jonas-Karl) [inline backend/docs/API_Documentation.yaml:181]:
```suggestion
```

delete line

File: backend/docs/API_Documentation.yaml [hunk 15]
```diff
@@ -112,173 +104,300 @@ paths:
...
-        $ref: '#/components/requestBodies/putCashflow'
+        $ref: '#/components/requestBodies/put_cashflow'
       responses:
         '200':
-          $ref: '#/components/responses/putAndDeleteCashflow.200'
+          $ref: '#/components/responses/put_and_delete_cashflow.200'
+
         '400':
           $ref: '#/components/responses/400'
+
```
Comment (Jonas-Karl) [inline backend/docs/API_Documentation.yaml:178]:
```suggestion
```

delete line

File: backend/docs/API_Documentation.yaml [hunk 16]
```diff
@@ -112,173 +104,300 @@ paths:
...
         - $ref: '#/components/parameters/auth'
         - $ref: '#/components/parameters/cashflow_id'
       requestBody:
-        $ref: '#/components/requestBodies/putCashflow'
+        $ref: '#/components/requestBodies/put_cashflow'
       responses:
         '200':
-          $ref: '#/components/responses/putAndDeleteCashflow.200'
+          $ref: '#/components/responses/put_and_delete_cashflow.200'
+
```
Comment (Jonas-Karl) [inline backend/docs/API_Documentation.yaml:175]:
```suggestion
```

delete line

File: backend/docs/API_Documentation.yaml [hunk 17]
```diff
@@ -112,173 +104,300 @@ paths:
...
         '400':
           $ref: '#/components/responses/400'
+
         '401':
-          $ref: '#/components/responses/accessError'
+          $ref: '#/components/responses/access_error'
+
         '500':
           $ref: '#/components/responses/cashflows.500'
+
```
Comment (Jonas-Karl) [inline backend/docs/API_Documentation.yaml:158]:
```suggestion
```

delete line

File: backend/docs/API_Documentation.yaml [hunk 18]
```diff
@@ -112,173 +104,300 @@ paths:
...
-          $ref: '#/components/responses/getCashflow.200'
+          $ref: '#/components/responses/get_cashflow.200'
+
         '400':
           $ref: '#/components/responses/400'
+
         '401':
-          $ref: '#/components/responses/accessError'
+          $ref: '#/components/responses/access_error'
+
```
Comment (Jonas-Karl) [inline backend/docs/API_Documentation.yaml:155]:
```suggestion
```

delete line

File: backend/docs/API_Documentation.yaml [hunk 19]
```diff
@@ -112,173 +104,300 @@ paths:
...
         - $ref: '#/components/parameters/auth'
         - $ref: '#/components/parameters/cashflow_id'
       responses:
         '200':
-          $ref: '#/components/responses/getCashflow.200'
+          $ref: '#/components/responses/get_cashflow.200'
+
         '400':
           $ref: '#/components/responses/400'
+
```
Comment (Jonas-Karl) [inline backend/docs/API_Documentation.yaml:152]:
```suggestion
```

delete line

File: backend/docs/API_Documentation.yaml [hunk 20]
```diff
@@ -112,173 +104,300 @@ paths:
...
       security:
         - bearerAuth: [ ]
       parameters:
         - $ref: '#/components/parameters/auth'
         - $ref: '#/components/parameters/cashflow_id'
       responses:
         '200':
-          $ref: '#/components/responses/getCashflow.200'
+          $ref: '#/components/responses/get_cashflow.200'
+
```
Comment (Jonas-Karl) [inline backend/docs/API_Documentation.yaml:149]:
```suggestion
```

delete line

File: backend/docs/API_Documentation.yaml [hunk 21]
```diff
@@ -112,173 +104,300 @@ paths:
...
         - bearerAuth: [ ]
       parameters:
         - $ref: '#/components/parameters/auth'
       requestBody:
-        $ref: '#/components/requestBodies/postPlace'
+        $ref: '#/components/requestBodies/post_store'
+      responses:
+        '200':
+          $ref: '#/components/responses/post_store'
+
```
Comment (Jonas-Karl) [inline backend/docs/API_Documentation.yaml:222]:
```suggestion
```

delete line

File: backend/docs/API_Documentation.yaml [hunk 22]
```diff
@@ -112,173 +104,300 @@ paths:
...
+        '400':
+          $ref: '#/components/responses/400'
+
```
Comment (Jonas-Karl) [inline backend/docs/API_Documentation.yaml:225]:
```suggestion
```

delete line

File: backend/docs/API_Documentation.yaml [hunk 23]
```diff
@@ -112,173 +104,300 @@ paths:
...
+        '401':
+          $ref: '#/components/responses/access_error'
+
```
Comment (Jonas-Karl) [inline backend/docs/API_Documentation.yaml:228]:
```suggestion
```

delete line

File: backend/docs/API_Documentation.yaml [hunk 24]
```diff
@@ -112,173 +104,300 @@ paths:
...
+      security:
+        - bearerAuth: [ ]
+      parameters:
+        - $ref: '#/components/parameters/auth'
+        - $ref: '#/components/parameters/private_id'
       responses:
         '200':
-          $ref: '#/components/responses/postPlace'
+          $ref: '#/components/responses/get_private'
+
```
Comment (Jonas-Karl) [inline backend/docs/API_Documentation.yaml:289]:
```suggestion
```

delete line

File: backend/docs/API_Documentation.yaml [hunk 25]
```diff
@@ -112,173 +104,300 @@ paths:
...
         '400':
           $ref: '#/components/responses/400'
+
```
Comment (Jonas-Karl) [inline backend/docs/API_Documentation.yaml:292]:
```suggestion
```

delete line

File: backend/docs/API_Documentation.yaml [hunk 26]
```diff
@@ -112,173 +104,300 @@ paths:
...
         '401':
-          $ref: '#/components/responses/accessError'
+          $ref: '#/components/responses/access_error'
+
```
Comment (Jonas-Karl) [inline backend/docs/API_Documentation.yaml:295]:
```suggestion
```

delete line

File: backend/docs/API_Documentation.yaml [hunk 27]
```diff
@@ -350,51 +499,106 @@ components:
...
-          format: date
-          example: "2023-04-15 12:23:59.602445+00"
+          example: 'Shampoo'
+        value:
+          type: number
+          example: 1.99
+        amount:
+          type: integer
+          example: 1
+
```
Comment (Jonas-Karl) [inline backend/docs/API_Documentation.yaml:600]:
```suggestion
```

delete line

File: backend/docs/API_Documentation.yaml [hunk 28]
```diff
@@ -694,11 +862,44 @@ components:
...
+        example: 1
+    category_id:
+      name: 'category_id'
+      in: path
+      description: 'The `category_id` of the Category'
+      required: true
+      schema:
+        type: integer
+        example: 1
+
```
Comment (Jonas-Karl) [inline backend/docs/API_Documentation.yaml:902]:
```suggestion
```

delete line

File: backend/docs/postman_collection.json
Comment (Jonas-Karl) [inline backend/docs/postman_collection.json:1]: I will not be able to update this postman_collection anymore as there is no time left
